pallets/crowdloan-claim/src/lib.rs

```diff
@@ -566,195 +450,234 @@ pub mod pallet {
         ///
         /// Unsigned transactions are generally disallowed. However, since a contributor
         /// claiming a reward payout may not have the necessary tokens on the parachain to
-        /// pay the fees of the claim, the [`claim_reward`] transactions must be 
+        /// pay the fees of the claim, the [`claim_reward`] transactions must be
         /// unsigned.
         /// Here, we make sure such unsigned, and remember, feeless unsigned transactions
         /// can be used for malicious spams or Deny of Service (DoS) attacks.
         fn validate_unsigned(_source: TransactionSource, call: &Self::Call) -> TransactionValidity {
-
-            // Check if the claim transaction is not too speedy (avoid DoS/DDoS attacks).
-            let next_claim_transaction_tick = <ClaimTransactionTick<T>>::get();
-            if next_claim_transaction_tick > <frame_system::Pallet<T>>::block_number() {
-                return InvalidTransaction::Stale.into();
+            if let Call::claim_reward(
+                relaychain_account_id,
+                parachain_account_id,
+                claimed_amount,
+                identity_proof,
+                contribution_proof,
+            ) = call
+            {
+                // By checking the validity of the claim here, we ensure the extrinsic will not
+                // make it into a block (in case of a trusted node, not even into the pool)
+                // unless being valid. This is a trade-off between protecting the network from spam
+                // and paying validators for the work they are doing.
+                if !ProcessedClaims::<T>::contains_key(&relaychain_account_id)
+                    && Self::verify_contributor_identity_proof(
+                        relaychain_account_id.clone(),
+                        parachain_account_id.clone(),
+                        identity_proof.clone(),
+                    )
+                    .is_ok()
+                    && Self::verify_contribution_proof(
+                        relaychain_account_id.clone(),
+                        claimed_amount.clone(),
+                        contribution_proof.to_vec(),
+                    )
+                    .is_ok()
+                {
+                    // Only the claim reward transaction can be invoked via an unsigned regime
+                    return ValidTransaction::with_tag_prefix("CrowdloanClaimReward")
+                        .priority(T::ClaimTransactionPriority::get())
+                        .and_provides((relaychain_account_id, parachain_account_id, claimed_amount))
+                        .longevity(
+                            TryInto::<u64>::try_into(T::ClaimTransactionLongevity::get())
+                                .unwrap_or(64_u64),
+                        )
+                        .propagate(true)
+                        .build();
+                } else {
+                    return InvalidTransaction::BadProof.into();
```

Comment:
I think there can be `AlreadyClaimed` or `BadProof` depeneding on whiuch failed right?  maybe we can seperate those checks with respective errors ?

Reply:
Yes, good idea.